src/MDMTransaction.m

```diff
@@ -16,31 +16,108 @@
 
 #import "MDMTransaction.h"
 #import "MDMTransaction+Private.h"
+#import "MDMPerforming.h"
+#import "MDMPlan.h"
 
 @implementation MDMTransaction {
   NSMutableArray *_logs;
+  NSMutableSet *_namedPlans;
 }
 
 - (instancetype)init {
   self = [super init];
   if (self) {
     _logs = [NSMutableArray array];
+    _namedPlans = [NSMutableSet set];
   }
   return self;
 }
 
 - (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target {
-  MDMTransactionLog *log = [MDMTransactionLog new];
-  log.plans = @[ [plan copy] ];
-  log.target = target;
+  [self commonAddPlan:plan toTarget:target withName:nil];
+}
+
+- (void)addPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  [self removePlanNamed:name fromTarget:target];
+  [self commonAddPlan:plan toTarget:target withName:name];
+}
+
+- (void)commonAddPlan:(NSObject<MDMPlan> *)plan toTarget:(id)target withName:(NSString *)name {
+  MDMTransactionLog *log = [self newLogWithPlan:plan target:target name:name];
+  if (name.length) {
+    Class performerClass = [plan performerClass];
+    id performer = [[performerClass alloc] initWithTarget:target];
+    if ([performer conformsToProtocol:@protocol(MDMNamedPlanPerforming)] &&
+        [performer respondsToSelector:@selector(addPlan:withName:)]) {
+      [performer addPlan:plan withName:name];
+    }
+    [_namedPlans addObject:log];
+  }
   [_logs addObject:log];
 }
 
+- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
+  MDMTransactionLog *log = [self newLogWithPlan:nil target:target name:name];
+  if ([_namedPlans containsObject:log]) {
+    // need to get the log back from _namedPlans so we can get at the plans
+    MDMTransactionLog *retrievedLog = [_namedPlans member:log];
+    if ([target isEqual:retrievedLog.target]) {
+      for (id<MDMPlan>plan in [retrievedLog plans]) {
+        Class performerClass = [plan performerClass];
+        id performer = [[performerClass alloc] initWithTarget:[retrievedLog target]];
+        if ([performer conformsToProtocol:@protocol(MDMNamedPlanPerforming)] &&
+            [performer respondsToSelector:@selector(removePlan:withName:)]) {
+          [performer removePlan:plan withName:name];
+        }
+      }
+    }
+    [_namedPlans removeObject:retrievedLog];
+    [_logs removeObject:retrievedLog];
```

Comment:
Been trying a few things in order to try to hit the `removePlanNamed:` callback with some unit tests, but after moving the logic to `MDMPerformerGroup`, I'm having difficulty seeing the callback fire.

I'm assuming that this is mostly due to this `removeObject` call:
```
- (void)removePlanNamed:(nonnull NSString *)name fromTarget:(nonnull id)target {
  MDMTransactionLog *log = [[MDMTransactionLog alloc] initWithPlan:nil target:target name:name transactionLogType:MDMTransactionLogTypeRemoveNamedPlan];
  [_logs removeObject:log];
}
```
Basically, this ensures that the scheduler will never see any `removePlan` `MDMTransactionLog`? I was mostly basing this code off https://github.com/material-motion/material-motion-runtime-android/blob/develop/library/src/main/java/com/google/android/material/motion/runtime/Transaction.java#L80.

I'm beginning to think that I should be adding the removePlan transaction into the `_logs` collection instead of just removing it?

https://codecov.io/gh/material-motion/material-motion-runtime-objc/compare/2adef827e4077c164e70631d0be1eb4ea8258a2e...29880f066ef3ae7a83a6c4fefd1f87a398bfef91 is the link for the code coverage for these few tests FWIW.